ports/mimxrt/machine_sdcard.c

```diff
@@ -0,0 +1,222 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#if MICROPY_PY_MACHINE_SDCARD
+
+#include "py/runtime.h"
+#include "py/mperrno.h"
+#include "extmod/vfs.h"
+#include "ticks.h"
+#include "fsl_cache.h"
+
+#include "sdcard.h"
+
+
+enum { SDCARD_INIT_ARG_ID };
+
+
+STATIC const mp_arg_t sdcard_init_allowed_args[] = {
+    [SDCARD_INIT_ARG_ID] = { MP_QSTR_id, MP_ARG_INT, {.u_int = 1} },
+};
+
+
+STATIC void machine_sdcard_init_helper(mimxrt_sdcard_obj_t *self) {
+    sdcard_init(self, 198000000UL);  // Initialize SDCard Host with 198MHz base clock
+    sdcard_init_pins(self);
+
+    ticks_delay_us64(2ULL * 1000ULL);  // Wait 2ms to allow USDHC signals to settle/debounce
+}
+
+STATIC mp_obj_t sdcard_obj_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    mp_map_t kw_args;
+    mp_map_init_fixed_table(&kw_args, n_kw, all_args + n_args);
+
+    // Parse args
+    mp_arg_val_t args[MP_ARRAY_SIZE(sdcard_init_allowed_args)];
+    mp_arg_parse_all(n_args, all_args, &kw_args, MP_ARRAY_SIZE(sdcard_init_allowed_args), sdcard_init_allowed_args, args);
+
+    // Extract arguments
+    mp_int_t sdcard_id = args[SDCARD_INIT_ARG_ID].u_int;
+
+    if (!(1 <= sdcard_id && sdcard_id <= MP_ARRAY_SIZE(mimxrt_sdcard_objs))) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError, "SDCard(%d) doesn't exist", sdcard_id));
+    }
+
+    mimxrt_sdcard_obj_t *self = &mimxrt_sdcard_objs[(sdcard_id - 1)];
+
+    // Initialize SDCard Host
+    if (!sdcard_state_initialized(self)) {
+        machine_sdcard_init_helper(self);
+    }
+    return MP_OBJ_FROM_PTR(self);
+}
+
+// init()
+STATIC mp_obj_t machine_sdcard_init(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    mimxrt_sdcard_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+
+    if (!sdcard_state_initialized(self)) {
+        machine_sdcard_init_helper(self);
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(machine_sdcard_init_obj, 1, machine_sdcard_init);
+
+// deinit()
+STATIC mp_obj_t machine_sdcard_deinit(mp_obj_t self_in) {
+    mimxrt_sdcard_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    sdcard_deinit(self);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_sdcard_deinit_obj, machine_sdcard_deinit);
+
+// readblocks(block_num, buf)
+STATIC mp_obj_t machine_sdcard_readblocks(mp_obj_t self_in, mp_obj_t block_num, mp_obj_t buf) {
+    mp_buffer_info_t bufinfo;
+    mimxrt_sdcard_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_WRITE);
+
+    if (sdcard_read(self, bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / SDCARD_DEFAULT_BLOCK_SIZE)) {
+        return MP_OBJ_NEW_SMALL_INT(0);
+    } else {
+        return MP_OBJ_NEW_SMALL_INT(-1);
```

Comment:
-1 would mean EPERM.  Would be better to return `-MP_EIO` (or something more specific like `MP_ETIMEDOUT` if `sdcard_read` knows of a more specific error)

Reply:
fixed. Changed all to`-E_IO` since that makes the most sense.